Improve crafting recipe documentation

Wuzzy2 opened the pull request:
Fixes #12796.

This adds missing documentation for the crafting recipe definition in `lua_api.txt`. Currently, the "documentation" was mostly just examples, not a proper definition. So I have written the missing definition, and parameter specification, and stuff.

Just read the diff for details. :P

Wuzzy2 commented:
Update! I have clarified some stuff about the crafting `replacements` and went into great detail, because those are pretty tricky.

TurkeyMcMac reviewed (approved):
LGTM

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
-Used by `minetest.register_craft`.
+All crafting recipes take one or more items as input and have an
+optional output.
+Crafting needs inventory slots. From the crafting inputs, only one
```
> Crafting needs inventory slots. 

What do you mean by that? `minetest.get_craft_result` for example does not need inventory slots, and the commonly known act of "crafting" obviously needs space to work with.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
 ----------------
 
-Used by `minetest.register_craft`.
+All crafting recipes take one or more items as input and have an
+optional output.
+Crafting needs inventory slots. From the crafting inputs, only one
+item per slot can be used per crafting, while the output is a single
+item stack of arbitrary size.
```
```suggestion
Crafting converts one or more inputs to one output ItemStack of arbitrary count
(including empty). The conversion reduces each input ItemStack by one (1).
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
+item per slot can be used per crafting, while the output is a single
+item stack of arbitrary size.
+
+Crafting recipe definitions are always a table with a number of
+parameters (listed below). They must be used as input parameter
+for `minetest.register_craft`.
```
```suggestion
Craft recipes are registered by `minetest.register_craft`. The accepted
formats are listed below..
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
+parameters (listed below). They must be used as input parameter
+for `minetest.register_craft`.
+
+Input items can be specified via itemname, that is, without itemcount
+or wear. (The item count is always 1.)
+
+Input slots may alternatively accept an item from a group, with a syntax like
+`group:<group_name>`. See "Groups in crafting recipes" for details.
```
```suggestion
Recipe input items can either be specified by item name (item count = 1)
or by group (see "Groups in crafting recipes" for details).
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
+For example, for a 3x3 recipe, the `recipes` table must have
+3 rows and 3 columns.
+
+For the player to be able to craft a shaped recipe, they must
+have access to a crafting grid that has the same or larger
+width and height as the recipe. E.g. a crafting grid of
+size 2x2 can only craft recipes of size 2x2, 2x1, 1x2 and 1x1,
+but not recipes of size 3x3, 3x2, 2x3, etc.
```
This largely seems superfluous/obvious.
```suggestion
In order to craft the recipe, the players' crafting grid must have matching or larger dimensions.
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
+Parameters:
+
+* `type = "shaped"`: (optional) specifies recipe type as shaped
+* `output`: Itemstring of output itemstack (item counts >= 1 are allowed)
```
```suggestion
* `output`: Itemstring of the output ItemStack
```
Crafting behaviour (count >= 1) is already documented above.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
+* `type = "shaped"`: (optional) specifies recipe type as shaped
+* `output`: Itemstring of output itemstack (item counts >= 1 are allowed)
+* `recipe`: A 2-dimensional matrix of items, with a width *w* and height *h*.
+    * *w* and *h* are chosen by you, they don't have to be equal but must be at least 1
+    * The matrix is specified as a table containing tables containing itemnames
+    * The inner tables are the rows. There must be *h* tables, specified from the top to the bottom row
+    * Values inside of the inner table are the columns.
+      Each inner table must contain a list of *w* items, specified from left to right
```
I suppose a commented code example would be more helpful than this text.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
+    * Values inside of the inner table are the columns.
+      Each inner table must contain a list of *w* items, specified from left to right
+    * Empty slots *must* be filled with the empty string
+* `replacements`: Optionally allows you to replace input items with some other items
+      when something is crafted
```
```suggestion
* `replacements`: (optional) replacement rules of input items on craft
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
+      `old_item` is the input item to replace (same syntax as for a regular input
+      slot; groups are allowed) and `new_item` is an itemstring for the item stack
+      it will become
+    * When the output is crafted, Minetest iterates through the list
```
What if the input stack has a count larger than 1? What's replaced?

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
+    -- Stone pickaxe
+    {
+        output = "example:stone_pickaxe",
+        -- This is a 3x3 recipe which needs 3 cobblestone in the 1st row,
+        -- and 1 stick in the horizontal middle in each of the 2nd and 3nd row.
+        -- The 4 remaining slots have to be empty.
+        recipe = {
+            {"example:cobblestone", "example:cobblestone", "example:cobblestone"},
+            {"", "example:stick", ""},
+            {"", "example:stick", ""},
```
```suggestion
        -- A 3x3 recipe containing 4 empty slots.
       recipe = {
            {"mymod:stone", "mymod:stone", "mymod:stone"}, -- row 1
            {"",            "mymod:stick", ""           }, -- row 2
            {"",            "mymod:stick", ""           }, -- row 3
            -- ^ column 1   ^ column 2     ^ column 3
          
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
 
 ### Shapeless
 
+Same as shaped recipe, except the arrangement of items in the input
+crafting grid does not matter. Also, the recipe is a list of items instead
+of a 2D matrix.
```
```suggestion
Takes an 1-dimensional array of at least one input item. The order
of input items does not matter.
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -8474,44 +8493,198 @@ Used by `minetest.register_node`.
+crafting grid does not matter. Also, the recipe is a list of items instead
+of a 2D matrix.
+
+For the player to be able to craft a shapeless recipe, they must have
+access to a crafting grid that has at least as many slots
+as the number of required items. E.g. a crafting grid of size 3x3 can
+craft shapeless recipes with up to 9 input items. The width of height of
+the grid do *not* matter.
```
This largely seems superfluous/obvious.
```suggestion
In order to craft the recipe, the players' crafting grid must have matching or
larger *count* of slots. The grid dimensions do not matter.
```

SmallJoker reviewed (commented):
Some suggestions to shorten the text while keeping the important information.

Wuzzy2 commented:
I have accepted and applied most of these changes (new commit), except:

> Crafting converts one or more inputs to one output ItemStack of arbitrary count
> (including empty). The conversion reduces each input ItemStack by one (1).

This is wrong or misleading, the output can NOT be specified as an ItemStack because ItemStack is an object. I changed it to "item stack" that term doesn't imply an object.
The output is only empty for fuel recipes. Otherwise, output is required and Minetest even refuses to start if you have a non-fuel
recipe with empty output. I didn't explain that correctly tho.

> Craft recipes are registered by `minetest.register_craft`. The accepted
> formats are listed below..

The fact that it's a table still should be mentioned explicitly.

> > optional output
>
> Is that actually possible? I haven't seen this being used anywhere. 

Fuel recipes never have an output. I changed the text anyway because this was confusing.

> What if the input stack has a count larger than 1? What's replaced?

I now explained this is more detail in a new bullet point.

> [the first example]

I intentionally want to keep a long explanation for the first example. That's the point of having examples.

Wuzzy2 commented:
More reactions because I overlooked these:

> * `output`: Itemstring of the output ItemStack
>
> Crafting behaviour (count >= 1) is already documented above.

Since this is the parameter list, I think it is OK to repeat it once. Also, ItemStack is misleading as it must not be an actual ItemStack object.

> I suppose a commented code example would be more helpful than this text.

Replacing documentation by examples is exactly what documentation should NOT be doing. This was a direct reaction from a complaint in the issue this PR is reacting to. Also, examples are already provided. I consider this to be irreducible complexity.

SmallJoker commented:
> This is wrong or misleading, the output can NOT be specified as an ItemStack because ItemStack is an object.

Interesting. ItemStacks are not accepted but any item string works, including tool wear - likely even with metadata. Weird but not related to PR :)

> Replacing documentation by examples

I did not mean to replace the documentation, but to integrate it into a generic example to comprehend what's going on while seeing the table format. I wouldn't read the text and head on to the examples, but other people might act different. Fair enough.

SmallJoker reviewed (approved):
Looks about right.

runsy commented:
Unfortunately 2x2 recipes are broken

TurkeyMcMac commented:
Could you open an issue?

runsy commented:
> Could you open an issue?

Yep
